pylint/checkers/variables.py

```diff
@@ -728,23 +733,43 @@ def _exhaustively_define_name_raise_or_return(
         if NamesConsumer._defines_name_raises_or_returns(name, node):
             return True
 
-        # If there is no else, then there is no collectively exhaustive set of paths
-        if not node.orelse:
-            return False
-
+        test = node.test.value if isinstance(node.test, nodes.NamedExpr) else node.test
+        all_inferred = utils.infer_all(test)
+        # Only search else branch when test condition is inferred to be false
+        if all_inferred and all(
+            isinstance(inferred, nodes.Const) and not inferred.value
+            for inferred in all_inferred
+        ):
+            return NamesConsumer._branch_handles_name(name, node.orelse)
+        # Only search if branch when test condition is inferred to be true
+        if all_inferred and any(
+            isinstance(inferred, nodes.Const)
+            and inferred.value != NotImplemented
+            and inferred.value
+            for inferred in all_inferred
+        ):
+            return NamesConsumer._branch_handles_name(name, node.body)
```

Comment:
This is a neat solution, but can we iterate `all_inferred` only once to determine the correct call to make?

Reply:
sure, i updated it to iterate over `all_inferred` once